crates/state/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+    }
+
+    pub async fn commit(&mut self) -> Result<Commit> {
+        let commit = self.checkpoint().await?;
+        self.next_commit_id += 1;
+        self.create_commit(commit).await?;
+        Ok(commit)
+    }
+
+    /// Persists all dirty changes from memory to disk. Even though persisted to
+    /// disk, they still don't belong to any commit; look into
+    /// [`GlobalState::commit`].
+    pub async fn checkpoint(&mut self) -> Result<Commit> {
+        let dirty_changes = std::mem::take(&mut self.dirty_changes);
+
+        let mut signature = [0; 32];
+
+        for change in dirty_changes {
+            let old_value: Option<Vec<u8>> = self.get_by_hash(&change.0, None).await?;
+            if let Some(old_value) = old_value {
+                xor_signature(&mut signature, &hash_key_value_pair(&change.0, &old_value));
+            }
+            xor_signature(&mut signature, &hash_key_value_pair(&change.0, &change.1));
+
+            let key_hash = &change.0[..];
+            let value = &change.1[..];
+            sqlx::query!(
+                r#"
+                    INSERT INTO "values" ("commit_id", "key_hash", "value")
+                    VALUES (?1, ?2, ?3)
+                "#,
+                self.next_commit_id,
+                key_hash,
+                value
+            )
+            .execute(&self.sqlite)
+            .await?;
+        }
+
+        self.update_commit_signature(&signature).await
+    }
+
+    async fn commit_signature(&self, commit_id: i64) -> Result<Commit> {
+        let bytes = sqlx::query!(
+            r#"
+                SELECT "signature"
+                FROM "commits"
+                WHERE "id" = ?1
+            "#,
+            commit_id
+        )
+        .fetch_one(&self.sqlite)
+        .await?
+        .signature;
+        Ok(bytes.try_into().unwrap())
+    }
+
+    async fn update_commit_signature(&mut self, partial_signature: &[u8; 32]) -> Result<Commit> {
+        let mut signature: [u8; 32] = self.commit_signature(self.next_commit_id).await?;
+        xor_signature(&mut signature, partial_signature);
+        let signature_bytes = &signature[..];
+
+        sqlx::query!(
+            r#"
+                UPDATE "commits"
+                SET "signature" = ?1
+                WHERE "id" = ?2
+            "#,
+            signature_bytes,
+            self.next_commit_id
+        )
+        .execute(&self.sqlite)
+        .await?;
+
+        Ok(signature)
+    }
+
+    /// Returns the [`Commit`] signature of the last ever checkpoint; i.e.
+    /// persisted changes without dirty changes.
+    pub async fn current(&self) -> Result<Commit> {
+        self.commit_signature(self.next_commit_id).await
+    }
+
+    /// Returns the current root hash in the form of a [`Commit`].
+    ///
+    /// It returns an error in case there's any dirty changes: you must
+    /// [`rollback`](GlobalState::rollback) first.
+    pub async fn rewind(&mut self, _commit: &Commit) -> Result<()> {
```

Comment:
I think that the rewind should do something like:

```rust
pub fn rewind(&mut self, commit_id: u64) {
   assert!(commit_id <= self.next_commit_id);

  self.next_commit_id = commit_id + 1;
}
```

We need the Global-State to be able to rewrite an existing `commit_id`.
If using the auto-increment feature collides with that requirement, maybe we can make it work manually.
(implementing the increment of the `commit_id` ourselves).

Since the client of the Global-State is known to work in a linear order I think it shouldn't be hard.

Reply:
The higher level API will probably have a similar:

```rust
pub fn rewind(&mut self, layer: u64) {
}
```

So a `layer` is essentially a `commit_id`

Reply:
Auto-incremental shouldn't cause any problems, I think we can leave it as it is! If you were to delete some commit ID, then you'd have a "hole". But that doesn't cause issues with range queries and similar.